sdk/rust/oak/src/handle.rs

```diff
@@ -0,0 +1,102 @@
+include!(concat!(env!("OUT_DIR"), "/oak.handle.rs"));
+
+use crate::OakError;
+
+pub type Handle = oak_abi::Handle;
+
+pub trait HandleVisit {
+    fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F;
```

Comment:
I find this type a bit awkward, especially since it seems it basically must always be used with a closure as parameter that actually accumulates or distributes values from the outside context. Have you considered instead something more similar to a [Foldable](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Foldable.html) / [Traversable](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Traversable.html) (I guess the closest in Rust would be https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold).

I haven't thought too much about this, but **I think** it may make the type something like

```rust
fn fold<B, F: FnMut(B, &mut Handle) -> B>(&mut self, init: B, f: F) -> B;
```

Reply:
Or maybe it could just implement some sort of iterator over handles? That may be sufficient to then extract and inject handles.

Reply:
You indeed always need a closure as a parameter to do anything meaningful with this, but I wouldn't personally consider that a problem, closures are easy to construct and quite common. The injection and extraction functions that use them also don't seem very complicated to me.

I think this could indeed also be implemented as a fold, then I **think** we could make the `FnMut` a `Fn`, since we don't need the closure itself to capture mutable variables. The price we'd pay for this is a slightly more complicated type signature and some loss of generality (which arguably we don't need). I'm not opposed to this idea but I don't see an advantage to it either.

An iterator *might* work too, but I think it would require significantly more code to implement (part of which has to be public because we need to derive it), and it does not make the implementation of `inject_handles` or `extract_handles` any simpler.

For reference, Serde uses a [visitor](https://docs.serde.rs/serde/de/trait.Visitor.html) pattern too, but instead of returning the closure after use, it has [special constructs](https://docs.serde.rs/serde/trait.Serializer.html#associatedtype.SerializeSeq) to handle recursion.

Reply:
Thanks, that makes sense, and anyways we can play around with types later on after this is merged, the tests will give us confidence that we are not doing something terrible hopefully :)